File: langs/typescript/src/commands/decoratorCommand.ts
/**
 * DecoratorCommand — wraps a single inner command with pre/post + extra-predicate.
 *
 * See spec/04-commands.md §Decorators and ADR-0012.
 */
import type { Observable } from "rxjs";
import type { ICommand } from "./types.js";

export interface DecoratorCommandOptions {
  preExecute?: () => void;
  postExecute?: () => void;
  extraPredicate?: () => boolean;
}

export class DecoratorCommand implements ICommand {
  readonly #inner: ICommand;
  readonly #pre: (() => void) | null;
  readonly #post: (() => void) | null;
  readonly #extra: (() => boolean) | null;

  constructor(inner: ICommand, opts: DecoratorCommandOptions = {}) {
    this.#inner = inner;
    this.#pre = opts.preExecute ?? null;
    this.#post = opts.postExecute ?? null;
    this.#extra = opts.extraPredicate ?? null;
  }

  get canExecuteChanged(): Observable<void> {
    return this.#inner.canExecuteChanged;
  }

  canExecute(): boolean {
    if (!this.#inner.canExecute()) return false;
    if (this.#extra === null) return true;
    try {
      return this.#extra();
    } catch {
      return false;
    }
  }

  execute(): void {
    if (!this.canExecute()) return;
    if (this.#pre) this.#pre();
    try {
      this.#inner.execute();
    } finally {
      // post runs whether or not the inner threw, so that a "busy" flag set
      // in preExecute always gets cleared.
      if (this.#post) this.#post();
    }
  }
}

File: langs/swift/Sources/VMx/Commands/DecoratorCommand.swift
//
// DecoratorCommand — wraps a single inner command with pre/post + extra-predicate.
//
// See spec/04-commands.md §8.2 and ADR-0012.
//
// Behaviour:
// - canExecute() = inner.canExecute() && (extraPredicate?() ?? true)
// - execute() when canExecute() is true: runs preExecute → inner.execute()
//   → postExecute, with postExecute in a `defer` block so it runs even if
//   inner.execute() were ever to throw (provides "busy-flag" teardown parity
//   with the TS/C# flavours whose inner can throw).
// - execute() is a complete no-op (pre/inner/post all skipped) when
//   canExecute() is false.
// - canExecuteChanged delegates to the inner command's publisher.
// - dispose() is a no-op: DecoratorCommand does not own the inner command.
//
import Foundation
import Combine

public final class DecoratorCommand: Command {
    private let inner: Command
    private let preExecute: (() -> Void)?
    private let postExecute: (() -> Void)?
    private let extraPredicate: (() -> Bool)?

    public init(
        _ inner: Command,
        preExecute: (() -> Void)? = nil,
        postExecute: (() -> Void)? = nil,
        extraPredicate: (() -> Bool)? = nil
    ) {
        self.inner = inner
        self.preExecute = preExecute
        self.postExecute = postExecute
        self.extraPredicate = extraPredicate
    }

    public func canExecute() -> Bool {
        guard inner.canExecute() else { return false }
        return extraPredicate?() ?? true
    }

    public func execute() {
        guard canExecute() else { return }
        preExecute?()
        defer { postExecute?() }
        inner.execute()
    }

    public var canExecuteChanged: AnyPublisher<Void, Never> {
        inner.canExecuteChanged
    }

    /// No-op: DecoratorCommand does not own the inner command and holds no
    /// subscriptions of its own. Provided for teardown symmetry with C#.
    public func dispose() {}
}
